daemon/cmd/config.go

```diff
@@ -51,29 +51,30 @@ func (h *patchConfig) Handle(params PatchConfigParams) middleware.Responder {
 
 	// Serialize configuration updates to the daemon.
 	option.Config.ConfigPatchMutex.Lock()
+	defer option.Config.ConfigPatchMutex.Unlock()
```

Comment:
These locking changes make me nervous. It looks very easy to misunderstand the locking patterns and introduce a deadlock. I've looked for about half an hour now and I can't convince myself it's correct (it may be; but even if it is, it looks easy to break in future).

I believe this is attempting to address the comments here:
https://github.com/cilium/cilium/pull/13971#discussion_r530021699

Furthermore, I think it is trying to address a case like:
1. API request comes in to configure some setting A
   1. Previous configuration is stored
   1. Configuration is applied
   1. Lock is released
   1. Reinitialization is triggered
1. API request comes in to configure some setting B
   1. Configuration is applied concurrently with reinitialization
   1. Lock is released
   1. Reinitialize is triggered
1. The step (1.iv) fails, so we need to undo the attempt at reconfiguration
   1. Overwrite the configuration with the original state, which undoes setting B modifications
1. The step (2.iii) succeeds, but (3) already reverted the changes

Clearly this is not the behaviour we want the API to have so I agree we should address it :+1: 

That said, I don't think it's a good idea to extend the locking over the entire `Reinitialize()` and try to enforce various places underneath to grab the right locks at the right time. This extends the locking over a long period, and also forces various locations to have more understanding of the locking because the locks aren't following a simple obvious "lock, read/change, unlock" pattern. In fact, the `Reinitialize()` path is shelling out to an external script which runs other processes like a compiler and so on, so we have very few guarantees about how long that will take. It could take seconds or even minutes in a CPU-starved environment (not typical, but possible).

The general way we solve this problem in other cases like policyAdd is to have an eventqueue where the events are serialized, then the reconfiguration is handled from there. If the caller wants to wait for the results, there is a channel-based mechanism to listen for the success or failure of the changes. See `putPolicy.Handle` and its callee `Daemon.PolicyAdd()` for examples.

Why is this different from just grabbing the lock? It allows serialization of the API requests without forcing the underlying data structures to be locked for a long period. The locking of the underlying structures, if necessary, can be handled briefly from various goroutines without introducing arbitrarily long delays. I believe this is already the case today for data like `IPv4NativeRoutingCIDR`, which currently has quite granular locking (which this PR is currently proposing to extend to a long-held lock).

With this approach in place, the core logic of the eventqueue's `struct ConfigModifyEvent` `Handle()` function could store the changes, lock, modify, unlock, `Reinitialize()`, then choose to either Lock, revert, Unlock or return success depending on the result of the reinitialization.

Originally I thought that maybe there's an alternative path, Undoing the changes in the event of a failure without holding the lock the whole time. For simple cases this *might* be doable to check that the settings are configured how we expect before reverting them in the error handling, but overall the most robust approach would be to follow the EventQueue approach.

Reply:
Thanks for the detailed write-up! I don't like the locking changes of various places as well. Sorry for not realizing the potential long execution time of `Reinitialize` and the benefit of using an event queue. Will try the suggested approach.